### DISCUSSION TITLE
ESLint to create a subset of JavaScript

### DISCUSSION THREAD

**Discussion by p8952** [eslint/eslint#17014] [Q&A]
I’m interested in creating a subset of JavaScript for education purposes. The idea being to allow the solving of problems using only specific parts of the language.

For example allowing the “.forEach()” and “.map()” methods on Arrays, but not allowing more “complex” methods like “.reduce()”.

Does anyone know if it would be possible to use ESLint as a basis for this?

Rather than writing rules to disallow certain methods, I’d like to essentially use ESLint in reverse and only allow methods I’ve explicitly written rules allowing.

**Comment by Rec0iL99** [comment]
Hi @p8952, you are probably looking for the [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule.

**Comment by nzakas** [comment]
Yes, this is possible, however it would be a pretty large effort to get down to blocking individual methods. Part of the problem is that you can't really know if a given identifier is an array or not, so all you can do is look for all method calls for "reduce" and infer that it's an array without knowing for sure.

**Comment by aladdin-add** [comment]
Basically, in the ESLint configuration, there is an option called "ecmaVersion" that allows you to specify the version of JavaScript to use. For example, "ecmaVersion: 5" means that only ES3/ES5 is used. In addition, there is a plugin `eslint-plugin-ex-x` that can disable certain features at a more granular level. e.g.

```js
rules: {'es-x/no-array-prototype-reduce': 'error'} // to disallow reduce
```

you can check the docs here: https://eslint-community.github.io/eslint-plugin-es-x/rules/no-array-prototype-reduce.html

**Reply by aladdin-add** [reply]
the rules may have some false positive/negative reports as lacking typing info (as @nzakas states :)).
